### PR TITLE
Graph view

### LINKED ISSUE #28: Display context graph
## User story
1. As a paying customer of CPU24/7 
2. I want / need to navigate through my data
3. So that I have an overview over interdependencies

## Acceptance criteria
* Favor existing open source solutions (keep it simple)
* Displayed as a subpage
* Visualized as a graph
* Being able to see all entities and the paths between them
* Integration of Filteroptions #47 

## Definition of done
* Added only after week 5
* The same for all features

### LINKED ISSUE #75: Simple graph frontend
Build a frontend for the backend implemented by #69 to complete the implementation of #28.

### PR DESCRIPTION
**Pull request by CatoLeanTruetschel**
This PR adds a rundimentary graph view as described by #75 in order to implement #28.
Closes #75
Closes #28

### CODE REVIEW

**Comment by CatoLeanTruetschel** [comment]
This now works but there is still some things to do
* Consistently get the base path of the backend, without hardcoding it. This would be significantly simplified with #60 
* Move the entities and helper function that are used in the frontend and in the backend to the shared folder
* Move the entity post processing from the frontend to the backend
* The resize of the window does not trigger a graph redraw reliably

File: frontend/src/@types/react-graph-vis.d.ts [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      identifier: string;
+    }
+  
+    export default class NetworkGraph extends Component<
+      NetworkGraphProps,
+      NetworkGraphState
+    > {
+      render();
+    }
+  }
```
Comment (joluj) [inline frontend/src/@types/react-graph-vis.d.ts:37]: no new line

File: frontend/src/entities/Edge.ts [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+/**
+ * Represents an edge of a graph with additional information.
+ */
+export interface Edge extends EdgeDescriptor {
+  type: string;
+  properties: {
+    [key: string]: Property;
+  };
+}
```
Comment (joluj) [inline frontend/src/entities/Edge.ts:13]: no new line

File: frontend/src/entities/EdgeDescriptor.ts [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    id: number;
+    /**
+     * ID of the start node
+     */
+    from: number;
+    /**
+     * ID of the end node
+     */
+    to: number;
+  }
```
Comment (joluj) [inline frontend/src/entities/EdgeDescriptor.ts:15]: no new line

File: frontend/src/routing/ListItemLink.tsx [hunk 4]
```diff
@@ -0,0 +1,44 @@
+import React from 'react';
+import { NavLink, LinkProps } from 'react-router-dom';
+import ListItem from '@material-ui/core/ListItem';
+import ListItemIcon from '@material-ui/core/ListItemIcon';
+import ListItemText from '@material-ui/core/ListItemText';
+
+interface ListItemLinkProps {
```
Comment (joluj) [inline frontend/src/routing/ListItemLink.tsx:7]: no comment

File: frontend/src/services/QueryService.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { injectable } from 'inversify';
+import 'reflect-metadata';
+import { LimitQuery } from '../entities/queries/LimitQuery';
+import { QueryResult } from '../entities/queries/QueryResult';
+import { CancellationToken } from '../utils/CancellationToken';
+
+@injectable()
```
Comment (joluj) [inline frontend/src/services/QueryService.ts:7]: no comments, especially for the parameters of the method

File: frontend/src/services/QueryServiceImpl.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+export default class QueryServiceImpl extends QueryService {
+  private readonly baseUri = 'http://localhost:8080'; // TODO: Make me configurable
+
+  public queryAll(
+    query?: LimitQuery,
+    cancellation?: CancellationToken
+  ): Promise<QueryResult> {
+    const url = `${this.baseUri}/queryAll`;
+
+    return new Promise<QueryResult>((resolve, reject) => {
```
Comment (joluj) [inline frontend/src/services/QueryServiceImpl.ts:20]: Can we put all that "http-stuff" with cancel events etc. in a new class? We surely need this functionality in other methods as well.
But its good enough ATM.

**Review by joluj — CHANGES_REQUESTED** [review]
Without any comments it's really hard to see whats going on.. The code looks good though.
The graph is shown, with the flaws you mentioned.

When the Code is documented then it's ready for the merge.

File: frontend/src/services/QueryServiceImpl.ts (hunk 6 above)
Comment (CatoLeanTruetschel) [reply frontend/src/services/QueryServiceImpl.ts:20]: Had this in mind, but as this is just needed in this single place currently did not invest the effort to actually factor it out yet.

File: frontend/src/@types/react-graph-vis.d.ts (hunk 1 above)
Comment (CatoLeanTruetschel) [reply frontend/src/@types/react-graph-vis.d.ts]: Done

File: frontend/src/entities/Edge.ts (hunk 2 above)
Comment (CatoLeanTruetschel) [reply frontend/src/entities/Edge.ts]: This file does no longer exist.

File: frontend/src/entities/EdgeDescriptor.ts (hunk 3 above)
Comment (CatoLeanTruetschel) [reply frontend/src/entities/EdgeDescriptor.ts]: This file does no longer exist.

File: frontend/src/routing/ListItemLink.tsx (hunk 4 above)
Comment (CatoLeanTruetschel) [reply frontend/src/routing/ListItemLink.tsx]: This file was not created, nor changed by this PR

**Comment by CatoLeanTruetschel** [comment]
All TODOs are resolved by now except for
> The resize of the window does not trigger a graph redraw reliably

**Review by PianoRollRepresentation — APPROVED** [review]

**Pull request merged**